ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -226,13 +227,16 @@ initNodeKernel args@NodeKernelArgs { registry, cfg, tracers
               , GSM.setCaughtUpPersistentMark = \upd ->
                   (if upd then GSM.touchMarkerFile else GSM.removeMarkerFile)
                     gsmMarkerFileView
-              , GSM.writeGsmState = \x -> atomically $ do
-                  writeTVar varLedgerJudgement $ GSM.gsmStateToLedgerJudgement x
+              , GSM.writeGsmState = \gsmState -> do
+                  handles <- atomically $
+                    writeTVar varGsmState gsmState >>
+                    readTVar varChainSyncHandles
+                  traverse_ (($ gsmState) . cschGsmCallback) handles
```

Comment:
Consider the scenario where we disconnect from a peer in-between fetching the handles  and executing the GSM callbacks. Wouldn't that mean that we might start the bucket background thread again? Doesn't seem super severe, but still a bit weird.

(Also, we don't seem to have any tests for `updateConfig`, and the implementation doesn't seem that thread-safe/async-exception-safe at a cursory glance. Maybe a `runSimStrictShutdown` test could have caught this?)

Reply:
> Consider the scenario where we disconnect from a peer in-between fetching the handles and executing the GSM callbacks. Wouldn't that mean that we might start the bucket background thread again? Doesn't seem super severe, but still a bit weird.

I agree both that it wouldn't be the worst but that it's weird. Yet another place where I'd want everything to happen in STM or in something thread-safe.

> (Also, we don't seem to have any tests for updateConfig, and the implementation doesn't seem that thread-safe/async-exception-safe at a cursory glance. Maybe a runSimStrictShutdown test could have caught this?)

Oh, I was sure there was such a test. Let me fix that.

Reply:
You're right; definitely not thread-safe, and the tests in f4539402d418906b1c65e4c6c81257592be6be3e fail precisely because of races.

Reply:
Fixed in https://github.com/IntersectMBO/ouroboros-consensus/pull/1007/commits/dc6ca25347f2984df61e495056e3983764aa1b23 à la https://github.com/IntersectMBO/ouroboros-consensus/pull/1007#issuecomment-2070080993.